### PR TITLE
Feature/distributed charging

### LINKED ISSUE #72: BusTool: add simulation output parameters
Is there more general simulation output parameters that we would like to have calculated by spice_EV? 
Then please add them to this list:

- [x] vehicles, which soc becomes negative
- [ ] ...

### PR DESCRIPTION
**Pull request by Piranias**
Fix #72, #41

**Changes proposed in this pull request**:
- new strategy: distributed with balanced strategy for gc's with "depot" ending and greedy strategy with gc's with "opp" ending
- restructure scenario.py and generate_schedule.py for more than one gc
- add tests

note: there is no generate_distributed_charging.py yet. The scenario.json will be generated within the bus-tool directly. Although there is an example bus_scenario.json in the test data.

The following steps were realized, as well (required):
- [x] Correct linting with `fake8 path_to_script`
- [x] Check if tests pass locally (`pytest ./src/tests.py`)

Also the following steps were realized (if applies):
- [x] Write docstrings to your code
- [x] For new functionalities: Explain in readthedocs
- [x] Write test(s) for your new patch of code

*Please mark above checkboxes as following:*
- [ ] Open
- [x] Done

### CODE REVIEW

File: doc/source/charging_strategies.rst [hunk 1]
```diff
@@ -167,4 +167,34 @@ V2G supported: **YES**
...
     |                   |               | (see description above for options and explanations)    |
-    +-------------------+---------------+---------------------------------------------------------+
\ No newline at end of file
+    +-------------------+---------------+---------------------------------------------------------+
+
+Distributed
+===========
+
+Distributed is a strategy that supports multiple grid connectors. The ending of each charging station name indicates if it is a 'depot' or a 'opp' (opportunity charging) station. Vehicles connected to opp grid connectors are charged
+according to the 'greedy' strategy. Vehicles connected to depot grid connectors are charged according to the 'balanced' strategy. A maximum number of charging stations can be assigned for each grid connector ('number_cs').
```
Comment (stefansc1) [inline doc/source/charging_strategies.rst:176]: why is there a newline?

File: doc/source/charging_strategies.rst [hunk 2]
```diff
@@ -167,4 +167,34 @@ V2G supported: **YES**
...
+
+Prioritization of vehicles at stations with a limited number charging stations:
+
+If the number of charging stations is limited all vehicles that want to connect in this and future time steps (limited by C-HORIZON) are collected and ranked by their SoC. The vehicle(s) with lowest SoC are loaded first until their desired SoC is reached or the cehicle departs.
```
Comment (stefansc1) [inline doc/source/charging_strategies.rst:180]: is limited,
in this -> in the current
typo cehicle -> vehicle

File: generate_schedule.py [hunk 3]
```diff
@@ -13,25 +13,20 @@
 EPS = 1e-8
 
 
-def generate_flex_band(scenario, core_standing_time=None):
+def generate_flex_band(scenario, gc_index, core_standing_time=None):
```
Comment (stefansc1) [inline generate_schedule.py:16]: Use grid connector ID instead of index. This is easier, better to read and cleaner, as GC are saved in a dict and not in a list).

File: generate_schedule.py [hunk 4]
```diff
@@ -13,25 +13,20 @@
...
     """
-
-    assert len(scenario.constants.grid_connectors) == 1, "Only one grid connector supported"
-    gc = list(scenario.constants.grid_connectors.values())[0]
-
+    gc = list(scenario.constants.grid_connectors.values())[gc_index]
     # generate basic strategy
     s = strategy.Strategy(
-        scenario.constants, scenario.start_time, **{"interval": scenario.interval})
+        scenario.constants, scenario.start_time, **{"interval": scenario.interval, "margin": 1})
```
Comment (stefansc1) [inline generate_schedule.py:29]: margin = 1 solves the problem if cars leave at the same timestep they are arriving. It does not solve the different problem of not making the next trip because the SoC is too low (e.g., PHEV). Three options:

1. ignore (simulation may fail in post-processing)
2. include allow_negative_soc
3. instead of using margin=1, use a try...catch block around s.step(...) -> this would also catch all other simulation errors

File: generate_schedule.py [hunk 5]
```diff
@@ -78,7 +74,12 @@ def clamp_to_gc(power):
     }
 
     # get battery info: how much can be discharged in beginning, how much if fully charged?
-    batteries = s.world_state.batteries.values()
+    batteries = {}
+    for battery, values in s.world_state.batteries.items():
+        if s.world_state.batteries[battery].parent == list(
+                scenario.constants.grid_connectors)[gc_index]:
+            batteries.update({battery: values})
+    batteries = batteries.values()
```
Comment (stefansc1) [inline generate_schedule.py:82]: This is needlessly complicated. Suggestion: `batteries = [b for b in s.world_state.batteries.values() if b.parent == gcID]`

File: generate_schedule.py [hunk 6]
```diff
@@ -57,6 +52,7 @@ def clamp_to_gc(power):
             v2g_enabled = True
     average_efficiency /= total_vehicle_capacity
 
+    cars = {vid: [0, 0, 0] for vid in s.world_state.vehicles}
```
Comment (stefansc1) [inline generate_schedule.py:55]: why was this moved?

File: generate_schedule.py [hunk 7]
```diff
@@ -105,33 +105,34 @@ def clamp_to_gc(power):
...
-                        # try to understand this one
-                        dep = -((scenario.start_time - dep) // s.interval)
-                        factor = min((scenario.n_intervals - step_i) / (dep - step_i), 1)
-                        delta_soc *= factor
-                    vehicle_energy_needed = (delta_soc * v.battery.capacity) / v.battery.efficiency
-                    v.battery.soc = max(v.battery.soc, v.desired_soc)
-                    v2g = (v.battery.get_available_power(s.interval)
-                           * v.vehicle_type.v2g_power_factor) if v.vehicle_type.v2g else 0
-                    cars[vid] = [charging_power, vehicle_energy_needed, v2g]
+                if s.world_state.charging_stations[v.connected_charging_station].parent == \
```
Comment (stefansc1) [inline generate_schedule.py:116]: compare cs.parent with gcID. For better readability, I would declare the current cs at the beginning of the vehicle-loop:
```
for vid, v in ...:
  cs_id = v.connected_charging_station
  if cs_id is None:
    ...
  else:
    # vehicle at charging station: charge full, keep track of needed and available energy
    cs = s.world_state.charging...
    if cs.parent == gcID:
      ...
```

File: generate_schedule.py [hunk 8]
```diff
@@ -199,8 +199,10 @@ def generate_schedule(args):
 
     ts_per_hour = datetime.timedelta(hours=1) / s.interval
 
+    assert len(s.constants.grid_connectors) == 1, "Only one grid connector supported"
```
Comment (stefansc1) [inline generate_schedule.py:202]: how easy is it to extend the functionality to multiple GCs? This is something for the future, not for the current PR

File: generate_schedule.py [hunk 9]
```diff
@@ -199,8 +199,10 @@ def generate_schedule(args):
...
+
     # compute flexibility potential (min/max) for each timestep
-    flex = generate_flex_band(s, args.core_standing_time)
+    flex = generate_flex_band(s, gc_index=0, core_standing_time=args.core_standing_time)
```
Comment (stefansc1) [inline generate_schedule.py:205]: use gcID instead: `gcID = list(s.world_state.grid_connectors.keys())[0]`
Please define the gcID separately and not in the function call. Looks cleaner to me. Maybe adjust the comment above: `compute flexibility potential (min/max) of single grid connector for each timestep`

File: simulate.py [hunk 10]
```diff
@@ -15,7 +15,7 @@
     'inverse',
     'peak_load_window', 'flex_window',
     'schedule', 'schedule_foresight',
-    'v2g',
+    'v2g', 'distributed'
```
Comment (stefansc1) [inline simulate.py:18]: distributed should get its own line. Try to end lines with a comma so you can easily extend the list (makes it easier to see changes in the diff)
I wasn't even aware that inverse and v2g are still valid strategies? I think both can be removed.

File: src/events.py [hunk 11]
```diff
@@ -311,7 +311,6 @@ def __init__(self, obj):
         conversions = [
             ('estimated_time_of_arrival', util.datetime_from_isoformat),
             ('estimated_time_of_departure', util.datetime_from_isoformat),
-            ('connected_charging_station', str),
```
Comment (stefansc1) [inline src/events.py:314]: I think this is important? Why is this removed?

File: src/scenario.py [hunk 12]
```diff
@@ -68,17 +66,17 @@ def run(self, strategy_name, options):
 
         event_steps = self.events.get_event_steps(self.start_time, self.n_intervals, self.interval)
 
-        socs = []
-        costs = []
-        prices = []
+        socs = {gcID: [] for gcID in self.constants.grid_connectors.keys()}
+        costs = {gcID: [] for gcID in self.constants.grid_connectors.keys()}
+        prices = {gcID: [] for gcID in self.constants.grid_connectors.keys()}
```
Comment (stefansc1) [inline src/scenario.py:71]: save a shorthand to iterate over GCs: `gc_ids = self.constants.grid_connectors.keys()`
This will also come in useful later in the code.

File: src/scenario.py [hunk 13]
```diff
@@ -68,17 +66,17 @@ def run(self, strategy_name, options):
...
-        disconnect = []
-        feedInPower = []
+        extLoads = {gcID: [] for gcID in self.constants.grid_connectors.keys()}
+        totalLoad = {gcID: [] for gcID in self.constants.grid_connectors.keys()}
+        disconnect = {gcID: [] for gcID in self.constants.grid_connectors.keys()}
+        feedInPower = {gcID: [] for gcID in self.constants.grid_connectors.keys()}
         stepsPerHour = datetime.timedelta(hours=1) / self.interval
-        batteryLevels = {k: [] for k in self.constants.batteries.keys()}
-        connChargeByTS = []
+        batteryLevels = {gcID: {} for gcID in self.constants.grid_connectors.keys()}
```
Comment (stefansc1) [inline src/scenario.py:78]: batteryLevels should still be saved by battery, not GC. When plotting, you may display the parent of the battery, but the values should be saved by battery. As it is, this line will not work, as you later try to append to nothing (you'd need to initialize an empty array for each battery of each GC)

File: src/scenario.py [hunk 14]
```diff
@@ -167,255 +166,284 @@ def run(self, strategy_name, options):
...
+            sum_energy_per_window = {}
+            avg_stand_time = {}
+            avg_total_standing_time = {}
+            perc_stand_window = {}
+            total_car_cap = {}
+            total_car_energy = {}
+            avg_drawn = {}
+
+            for gc_index, gcID in enumerate(self.constants.grid_connectors):
+                flex = generate_flex_band(self, gc_index=gc_index)
```
Comment (stefansc1) [inline src/scenario.py:244]: `generate_flex_band(self, gcID = gcID)`

File: src/scenario.py [hunk 15]
```diff
@@ -425,7 +453,9 @@ def run(self, strategy_name, options):
             if ext != "csv":
                 print("File extension mismatch: timeseries file is of type .csv")
 
-            cs_ids = sorted(strat.world_state.charging_stations.keys())
+            cs_ids = sorted(item for item in strat.world_state.charging_stations.keys() if
+                            self.constants.charging_stations[item].parent == gcID)
```
Comment (stefansc1) [inline src/scenario.py:457]: what is gcID? I think it is the last GC looked at after the simulation. You'd need to have an array for each GC -> `cs_ids = {gcID: sorted(item ... if cs[item].parent == gcID) for gcID in gc_ids). This also goes for a lot of the lines after this. You have to iterate over the GCs, otherwise you only get one! Is it really your intention to create output files for each GC individually?

File: src/scenario.py [hunk 16]
```diff
@@ -167,255 +166,284 @@ def run(self, strategy_name, options):
...
+                        "unit": None,
+                        "info": "Number of load cycles per vehicle (averaged)"
+                    }
+
+                    if options.get("save_results", False):
+                        # write to file
+                        file_name = options['save_results'].split(".")[0] + f"_{gcID}." + options[
+                            'save_results'].split(".")[1]
+                        with open(file_name, 'w') as results_file:
+                            json.dump(json_results, results_file, indent=2)
```
Comment (stefansc1) [inline src/scenario.py:446]: assuming there is only one gc, how does the output differ from before? Can the old evaluation tools still be used?

File: src/strategies/balanced.py [hunk 17]
```diff
@@ -116,46 +65,161 @@ def step(self, event_list=[]):
...
+            power = -gc.get_current_load()
+            power = 0 if power < battery.min_charging_power else power
+            avg_power = battery.load(self.interval, power)['avg_power']
+            gc.add_load(b_id, avg_power)
+        else:
+            # GC draws power: use stored energy to support GC
+            bat_power = battery.unload(self.interval,
+                                       gc.get_current_load() / battery.efficiency
+                                       )['avg_power']
+            gc.add_load(b_id, -bat_power)
```
Comment (stefansc1) [inline src/strategies/balanced.py:225]: Use the functions from strategy.py instead. Also goes for greedy and distributed.

File: src/strategies/distributed.py [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+class Distributed(Strategy):
+    """
+    Strategy that allows for greedy charging at opp stops and balanced charging at depots.
+    """
+    def __init__(self, constants, start_time, **kwargs):
+        self.PRICE_THRESHOLD = 0.001  # EUR/kWh
+        super().__init__(constants, start_time, **kwargs)
+        self.description = "distributed"
+        self.ITERATIONS = 12
+        self.C_HORIZON = 3  # min
```
Comment (stefansc1) [inline src/strategies/distributed.py:18]: What is C_HORIZON? Comment missing

File: src/strategy.py [hunk 19]
```diff
@@ -31,9 +30,10 @@ def __init__(self, constants, start_time, **kwargs):
         self.world_state.future_events = []
         self.interval = kwargs.get('interval')  # required
         self.current_time = start_time - self.interval
+        self.count_negative_soc = {}
```
Comment (stefansc1) [inline src/strategy.py:33]: count_negative_soc not clear (comment missing). Seems more to keep track of times when SoC is negative than counting? -> rename to negative_soc_tracker?

File: src/strategy.py [hunk 20]
```diff
@@ -115,13 +114,16 @@ def step(self, event_list=[]):
                     assert vehicle.battery.soc >= (1-self.margin)*vehicle.desired_soc - self.EPS, (
                         "{}: Vehicle {} is below desired SOC ({} < {})".format(
                             ev.start_time.isoformat(), ev.vehicle_id,
-                            vehicle.battery.soc, vehicle.desired_soc))
+                            vehicle.battery.soc, vehicle.desired_soc - self.margin))
```
Comment (stefansc1) [inline src/strategy.py:117]: No. The error message should display the real desired SoC. Internally, the margin is applied, but if the desired SoC is not reached with the margin, it is definitely not reached without it.

**Review by stefansc1 — CHANGES_REQUESTED** [review]

File: src/events.py (hunk 11 above)
Comment (Piranias) [reply src/events.py:314]: no, this line was the problem why all "None" values were read in as a string (we hat this issue a while ago). The "connected_charging_stations" are already read in as a str.

File: src/scenario.py (hunk 16 above)
Comment (Piranias) [reply src/scenario.py:446]: the name of the file is changed to.g. "simulation_{gc_name}.json" and "simulation_{gc_name}.json"

File: src/scenario.py (hunk 15 above)
Comment (Piranias) [reply src/scenario.py:457]: yes, each gc has its own output file

File: src/strategies/balanced.py (hunk 17 above)
Comment (Piranias) [reply src/strategies/balanced.py:225]: these are not in dev yet, that's why I didn't use them..

File: src/scenario.py (hunk 13 above)
Comment (Piranias) [reply src/scenario.py:78]: thats how it is actually implemented right now, see line 216
